---
title: "compare_glmpca"
author: "Joel Eliason"
date: "5/18/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
library(tidyverse)
options(repos = BiocManager::repositories())
library(SingleCellExperiment)
# library(ensembldb)
library(scRNAseq)
# library(AnnotationHub)
library(scater)
library(scran)
library(scry)
```

In this notebook, I begin to compare the GLMPCA method for feature selection in single-cell RNA seq datasets with that of more traditional variance modeling and selection of high variance genes.

```{r, cache=TRUE}
# datasets = listDatasets()
# 
# sce <- MacoskoRetinaData()

path = '/media/joeleliason/Data/lung_timecourse/SCP971/other/ltraw.mat'

r = readMat(path)

counts = r$rawCount

colnames(counts) = unlist(r$sampleID, recursive = FALSE)
rownames(counts) = unlist(r$geneSymbol, recursive = FALSE)

# counts = counts[,-grep('_4w_',colnames(counts))]
sce = SingleCellExperiment(assays = list(counts = counts))

sce$typeID = do.call(c, unlist(r$typeID, recursive=FALSE))

spl = strsplit(sce$typeID,'_')
weeks = sapply(spl, function(x) x[3])
sce$weeks = weeks

# sce$geneID = do.call(c, unlist(r$geneID, recursive=FALSE))

rm(r,counts)
```

```{r, cache=TRUE}
sce = devianceFeatureSelection(sce, assay="counts", sorted = TRUE, batch = as.factor(sce$typeID))
```


```{r}
plot(rowData(sce)$binomial_deviance[1000:52638], type="l", xlab="ranked genes",
     ylab="binomial deviance", main="Feature Selection with Deviance")
abline(v=2000, lty=2, col="red")
```

```{r}
sce<-nullResiduals(sce, assay="counts", type="deviance")
sce<-nullResiduals(sce, assay="counts", type="pearson")
sce2<-sce[1:1000, ] #use only the high deviance genes
pca<-function(Y, L=2, center=TRUE, scale=TRUE){
    #assumes features=rows, observations=cols
    res<-prcomp(as.matrix(t(Y)), center=center, scale.=scale, rank.=L)
    factors<-as.data.frame(res$x)
    colnames(factors)<-paste0("dim", 1:L)
    factors
}
pca_d<-pca(assay(sce2, "binomial_deviance_residuals"))
pca_d$resid_type<-"deviance_residuals"
pca_p<-pca(assay(sce2, "binomial_pearson_residuals"))
pca_p$resid_type<-"pearson_residuals"
cm<-as.data.frame(colData(sce2))
pd<-rbind(cbind(cm, pca_d), cbind(cm, pca_p))
ggplot(pd, aes(x=dim1, y=dim2, colour=phenoid)) + geom_point() +
  facet_wrap(~resid_type, scales="free", nrow=2) +
  ggtitle("PCA applied to null residuals of high deviance genes")
```

